crates/ra_hir_def/src/item_tree.rs

```diff
@@ -0,0 +1,486 @@
+use hir_expand::{
+    ast_id_map::FileAstId,
+    hygiene::Hygiene,
+    name::{name, AsName, Name},
+};
+use ra_arena::{map::ArenaMap, Arena, Idx, RawId};
+use ra_syntax::{ast, AstPtr};
+
+use crate::{
+    attr::Attrs,
+    generics::GenericParams,
+    path::{path, GenericArgs, ImportAlias, ModPath, Path},
+    type_ref::{Mutability, TypeBound, TypeRef},
+    visibility::RawVisibility,
+};
+use ast::{NameOwner, StructKind, TypeAscriptionOwner};
+use either::Either;
+use rustc_hash::FxHashMap;
+use std::{ops::Range, sync::Arc};
+
+#[derive(Default)]
+pub struct ItemTree {
+    imports: Arena<Import>,
+    functions: Arena<Function>,
+    structs: Arena<Struct>,
+    fields: Arena<Field>,
+    unions: Arena<Union>,
+    enums: Arena<Enum>,
+    variants: Arena<Variant>,
+    consts: Arena<Const>,
+    statics: Arena<Static>,
+    traits: Arena<Trait>,
+    impls: Arena<Impl>,
+    type_aliases: Arena<TypeAlias>,
+    mods: Arena<Mod>,
+    macro_calls: Arena<MacroCall>,
+    exprs: Arena<Expr>,
+}
+
+#[derive(Default)]
+pub struct ItemTreeSrc {
+    functions: ArenaMap<Idx<Function>, AstPtr<ast::FnDef>>,
+    structs: ArenaMap<Idx<Struct>, AstPtr<ast::StructDef>>,
+    fields: ArenaMap<Idx<Field>, Either<AstPtr<ast::RecordFieldDef>, AstPtr<ast::TupleFieldDef>>>,
+    unions: ArenaMap<Idx<Union>, AstPtr<ast::UnionDef>>,
+    enums: ArenaMap<Idx<Enum>, AstPtr<ast::EnumDef>>,
+    variants: ArenaMap<Idx<Variant>, AstPtr<ast::EnumVariant>>,
+    consts: ArenaMap<Idx<Const>, AstPtr<ast::ConstDef>>,
+    statics: ArenaMap<Idx<Static>, AstPtr<ast::StaticDef>>,
+    traits: ArenaMap<Idx<Trait>, AstPtr<ast::TraitDef>>,
+    impls: ArenaMap<Idx<Impl>, AstPtr<ast::ImplDef>>,
+    type_aliases: ArenaMap<Idx<TypeAlias>, AstPtr<ast::TypeAliasDef>>,
+    mods: ArenaMap<Idx<Mod>, AstPtr<ast::Module>>,
+    macro_calls: ArenaMap<Idx<MacroCall>, AstPtr<ast::MacroCall>>,
+    exprs: ArenaMap<Idx<Expr>, AstPtr<ast::Expr>>,
+}
+
+impl ItemTree {
+    fn new(syntax: &ast::SourceFile) -> ItemTree {
+        ItemTree::default()
+    }
+}
+
+pub struct Import {
+    pub path: ModPath,
+    pub alias: Option<ImportAlias>,
+    pub visibility: RawVisibility,
+    pub is_glob: bool,
+    pub is_prelude: bool,
+    pub is_extern_crate: bool,
+    pub is_macro_use: bool,
+}
+
+pub struct Function {
```

Comment:
Shouldn't we story unsafety and asynchrony modifiers for `Function` and unsafety for `Trait`?